aten/src/ATen/native/BatchLinearAlgebraKernel.cpp

```diff
@@ -915,11 +914,30 @@ void apply_lu_factor(const Tensor& input, const Tensor& pivots, const Tensor& in
   auto n = input.size(-1);
   auto leading_dimension = std::max<int64_t>(1, m);
 
-  for (const auto i : c10::irange(batch_size)) {
-    scalar_t* input_working_ptr = &input_data[i * input_matrix_stride];
-    int* pivots_working_ptr = &pivots_data[i * pivots_stride];
-    int* infos_working_ptr = &infos_data[i];
-    lapackLu<scalar_t>(m, n, input_working_ptr, leading_dimension, pivots_working_ptr, infos_working_ptr);
+  std::function<void(int64_t, int64_t)> loop = [](int64_t, int64_t) {};
+  loop = [&](int64_t start, int64_t end) {
+    for (const auto i : c10::irange(start, end)) {
+      scalar_t* input_working_ptr = &input_data[i * input_matrix_stride];
+      int* pivots_working_ptr = &pivots_data[i * pivots_stride];
+      int* infos_working_ptr = &infos_data[i];
+      lapackLu<scalar_t>(
+          m,
+          n,
+          input_working_ptr,
+          leading_dimension,
+          pivots_working_ptr,
+          infos_working_ptr);
+    }
+  };
+  int64_t matrix_size = std::min(m, n);
+  int64_t computational_complexity =
+      batch_size * matrix_size * matrix_size * matrix_size;
+  // A threshold test out on 32 core/socket ICX system
+  bool do_parallel = computational_complexity >= 102400;
+  if (do_parallel) {
+    at::parallel_for(0, batch_size, 0, loop);
+  } else {
+    loop(0, batch_size);
   }
```

Comment:
Hmm, shouldn't we swap the order? Seems like the heavier the matrix, the better it is for LAPACK do things in parallel with serial batch loop, or am I missing something?

Reply:
Can we do parallel on the batch dim and devise a good grain size per batch to let `at::parallel_for` judge whether to do multi-threading or serial execution?

Reply:
Hi, @nikitaved and @jgong5.
The data shows while the matrix gets heavier(rank=64), parallel on small batch size (100) will still get a large speed up (>5x). So whether to use parallel is not only related to batch size but also related to matrix size. So I use the `computational_complexity` to decide whether use at::parallel outside. Maybe `computational_complexity size/current thread num` can reflect the speed up more precisely, but since there is almost a speed up on all thread num cases so we may not need to divide by thread num.

Reply:
What about matrices of sizes >= 1024, for example? I speculate that doing all-serial batch loop is better in this case, and cache is expected to be utilized more efficiently...

Reply:
How about devise the grain size according to the matrix size? The grain size is smaller when the matrix size is bigger. And we always do at::parallel_for on the batch dim.

> I speculate that doing all-serial batch loop is better in this case, and cache is expected to be utilized more efficiently... To generalize, it seems like there is a matrix size threshold past which it is always better to delegate parallelization to the linalg routine.

According to the PR description, it seems worse to do parallelization inside lapack: "Even for large matrix, the parallelization efficiency is not good in lapack ( it performs worse than using at::parallel outside)". Not sure why it is the case.

Reply:
The largest matrix in the benchmark is just 64 by 64.

Reply:
Hi, @nikitaved. I tried matrix size up to 32k * 32k and parallel in aten won't worse than parallel inside lapck.
| bs/rank | 32/1024 |  32/4096 | 32/8192 |  32/16384 |32/32768|
| ---- | --- | --- | --- | --- | --- | 
|speed up ratio| 2.75x | 1.23x|1.02x |1.02x |1.03x |

Reply:
> How about devise the grain size according to the matrix size? The grain size is smaller when the matrix size is bigger. And we always do at::parallel_for on the batch dim.
> 
> > I speculate that doing all-serial batch loop is better in this case, and cache is expected to be utilized more efficiently... To generalize, it seems like there is a matrix size threshold past which it is always better to delegate parallelization to the linalg routine.
> 
> According to the PR description, it seems worse to do parallelization inside lapack: "Even for large matrix, the parallelization efficiency is not good in lapack ( it performs worse than using at::parallel outside)". Not sure why it is the case.

Yes, now we are using matrix size to decide grain_size.

Reply:
@zhuhaozhe , the benchmarks are just for `float`, what about `double` and complex types?